narwhal/worker/src/batch_maker.rs

```diff
@@ -187,8 +204,30 @@ impl BatchMaker {
             .observe(size as f64);
 
         // Send the batch through the deliver channel for further processing.
-        if self.tx_message.send(batch).await.is_err() {
+        let (notify_done, done_sending) = tokio::sync::oneshot::channel();
+        if self
+            .tx_message
+            .send((batch.clone(), Some(notify_done)))
+            .await
+            .is_err()
+        {
             tracing::debug!("{}", DagError::ShuttingDown);
         }
+
+        // Now save it to disk
+        let digest = batch.digest();
+        self.store.write(digest, batch).await;
+        // Wait until the batch was written
+        let _ = self.store.notify_read(digest).await;
+
+        // Also wait for sending to be done here
+        // (TODO: sending to others is an optimization, no need to wait.)
+        let _ = done_sending.await;
```

Comment:
If we wait here, is it useful to run the quorum waiter on a separate task as the batch-maker?

Reply:
Now we wait while sending many batches in parallel, deprecating #4934.